pkg/ipam-controller/allocator/allocator.go

```diff
@@ -0,0 +1,373 @@
+	log := pa.getLog(ctx, pa.cfg).WithValues("node", node)
+	existingAlloc, exist := pa.allocations[node]
+	if exist {
+		log.V(1).Info("allocation for the node already exist",
+			"start", existingAlloc.StartIP, "end", existingAlloc.EndIP)
+		return pa.getNodeAllocationInfo(node, existingAlloc), nil
+	}
+	allocations := pa.getAllocationsAsSlice()
+	var startIP net.IP
+	if len(allocations) == 0 || ip.Distance(ip.Network(pa.cfg.Subnet).IP, allocations[0].StartIP) > 2 {
+		// start allocations from the network address if there are no allocations or if the "hole" exist before
+		// the firs allocation
+		startIP = ip.NextIP(ip.Network(pa.cfg.Subnet).IP)
+	} else {
+		for i := 0; i < len(allocations); i++ {
+			nextI := i + 1
+			// if last allocation in the list
+			if nextI > len(allocations)-1 ||
+				// or we found a "hole" in allocations. the "hole" can't be less than required for the allocation by design.
+				// because we reset all allocations when PerNodeBlockSize size changes
+				ip.Distance(allocations[i].EndIP, allocations[nextI].StartIP) > 1 {
+				startIP = ip.NextIP(allocations[i].EndIP)
+				break
+			}
+		}
+	}
+	nodeRange := allocatedRange{
+		StartIP: startIP,
+		EndIP:   ip.NextIPWithOffset(startIP, int64(pa.cfg.PerNodeBlockSize)-1),
+	}
+	if !pa.cfg.Subnet.Contains(nodeRange.EndIP) || ip.IsBroadcast(nodeRange.EndIP, pa.cfg.Subnet) {
+		// out of range
+		log.Info("can't allocate: pool has no free ranges")
+		return nodeAllocationInfo{}, ErrNoFreeRanges
+	}
+
+	pa.allocations[node] = nodeRange
+	log.Info("range allocated",
+		"start", nodeRange.StartIP, "end", nodeRange.EndIP)
+	return pa.getNodeAllocationInfo(node, nodeRange), nil
+}
+
+// Deallocate remove info about allocation for the node from the poolAllocator
+func (pa *poolAllocator) Deallocate(ctx context.Context, node string) {
+	log := pa.getLog(ctx, pa.cfg)
+	log.Info("deallocate range for node", "node", node)
+	delete(pa.allocations, node)
+}
+
+// Load loads range to the pool allocator with validation for conflicts
+func (pa *poolAllocator) Load(ctx context.Context, allocData nodeAllocationInfo) error {
+	log := pa.getLog(ctx, pa.cfg)
+	if err := pa.checkAllocation(allocData); err != nil {
+		log.Info("range check failed", "reason", err.Error())
+		return err
+	}
+	allocations := pa.getAllocationsAsSlice()
+	for _, a := range allocations {
+		if allocData.allocatedRange.StartIP.Equal(a.StartIP) {
+			err := fmt.Errorf("range overlaps with: %v", a)
+			log.Info("skip loading range", "reason", err.Error())
+			return err
+		}
+	}
+	pa.allocations[allocData.Node] = allocData.allocatedRange
+	return nil
+}
+
+func (pa *poolAllocator) checkAllocation(allocData nodeAllocationInfo) error {
+	if allocData.Subnet.String() != pa.cfg.Subnet.String() {
+		return fmt.Errorf("subnet mismatch")
+	}
+	if !allocData.Gateway.Equal(pa.cfg.Gateway) {
+		return fmt.Errorf("gateway mismatch")
+	}
+	if math.Mod(float64(ip.Distance(ip.Network(pa.cfg.Subnet).IP, allocData.StartIP))-1,
```

Comment:
-1 is because we skip the network address ? (e.g for 192.168.0.0/16 , 192.168.0.0 is not allocated)

Reply:
simple example:

network: 0.0.0.0
block size 10

StartIPs wil be: 1, 11, 21, 31, 41 ...
Distance per StartIP from Network IP is same
subsracting 1 will yield 0 in mod 10 (block size)

so it seems to work  :).